vendors/nordic/boards/nrf52840-dk/ports/common_io/iot_flash.c

```diff
@@ -0,0 +1,490 @@
+                                 size_t xSize )
+{
+    int32_t lReturnCode = IOT_FLASH_INVALID_VALUE;
+
+    if( !prvIsOperableHandle( pxFlashHandle ) || !prvIsAligned4K( ulStartAddress ) || ( xSize % FLASH_PAGE_SIZE ) )
+    {
+        lReturnCode = IOT_FLASH_INVALID_VALUE;
+    }
+    else if( nrf_fstorage_is_busy( &iot_fstorage ) )
+    {
+        lReturnCode = IOT_FLASH_DEVICE_BUSY;
+    }
+    else if( xSize == 0 )
+    {
+        lReturnCode = IOT_FLASH_SUCCESS;
+    }
+    else
+    {
+        ret_code_t xReturnCode_NRF = nrf_fstorage_erase( &iot_fstorage, ulStartAddress, xSize / FLASH_PAGE_SIZE, NULL );
+
+        switch( xReturnCode_NRF )
+        {
+            case NRF_SUCCESS:
+                prvWaitUntilFlashReady( &iot_fstorage );
+                lReturnCode = IOT_FLASH_SUCCESS;
+                break;
+
+            case NRF_ERROR_NO_MEM:
+                IotLogError( "%s: Insufficient memory to queue operation. Review sdk_config.h:FDS_OP_QUEUE_SIZE", __func__ );
+                lReturnCode = IOT_FLASH_ERASE_FAILED;
+                break;
+
+            case NRF_ERROR_INVALID_ADDR:
+                IotLogError( "%s: Out-of-bounds access (0x%x)", __func__, ulStartAddress );
+                lReturnCode = IOT_FLASH_INVALID_VALUE;
+                break;
+
+            default:
+                lReturnCode = IOT_FLASH_INVALID_VALUE;
+                break;
+        }
+    }
+
+    return lReturnCode;
+}
+
+
+int32_t iot_flash_erase_chip( IotFlashHandle_t const pxFlashHandle )
+{
+    return IOT_FLASH_FUNCTION_NOT_SUPPORTED;
+}
+
+
+int32_t iot_flash_write_sync( IotFlashHandle_t const pxFlashHandle,
+                              uint32_t ulAddress,
+                              uint8_t * const pvBuffer,
+                              size_t xBytes )
+{
+    int32_t lReturnCode = IOT_FLASH_INVALID_VALUE;
+
+    if( !prvIsOperableHandle( pxFlashHandle ) || ( pvBuffer == NULL ) || ( xBytes > MAX_WR_SIZE ) )
+    {
+        lReturnCode = IOT_FLASH_INVALID_VALUE;
+    }
+    else if( nrf_fstorage_is_busy( &iot_fstorage ) )
+    {
+        lReturnCode = IOT_FLASH_DEVICE_BUSY;
+    }
+    else
+    {
+        /* Prepare data for flash write. Must be word-multiple (4 B). Padded if necessary.
+         *
+         * Note: User can use eGetFlashTxNoOfbytes to read then form a word-multiple write,
+         *       and preserve any existing flash data that would otherwise be overwritten with pad-value
+         */
+        size_t write_size = 4 * ( ( xBytes + 3 ) / 4 ); /* pad size to multiple of a word */
+        uint8_t write_buf[ write_size ] __attribute__( ( aligned( 4 ) ) );
+        memcpy( write_buf, pvBuffer, xBytes );
+        memset( write_buf + xBytes, FLASH_PADDING_BYTE, write_size - xBytes );
+
+        /* Write and check status*/
+        ret_code_t xReturnCode_NRF = nrf_fstorage_write( &iot_fstorage, ulAddress, write_buf, write_size, NULL );
+
+        switch( xReturnCode_NRF )
```

Comment:
This error handler seems to be repeated in many places.  Do we think that a more "mature" version of the library will evolve to have different error responses for each of these functions?

Reply:
The handler is different for read_sync, though it looks similar. It is currently the same for write_sync and erase_sync. 
Exactly, left it as such in case we to further customize handler response for each case